File: sdap-typescript/src/crypto/encryption.ts
/**
 * AES-256-GCM encryption/decryption for SDAP session payloads.
 */

import { webcrypto } from "node:crypto";

function b64urlEncode(data: Uint8Array): string {
  return Buffer.from(data)
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=/g, "");
}

function b64urlDecode(s: string): Uint8Array {
  const padding = 4 - (s.length % 4);
  const padded = padding !== 4 ? s + "=".repeat(padding) : s;
  const b64 = padded.replace(/-/g, "+").replace(/_/g, "/");
  return new Uint8Array(Buffer.from(b64, "base64"));
}

function buildAad(
  sessionId: string,
  sequenceNumber: number,
  senderDid: string
): Uint8Array {
  const aadObj = {
    senderDID: senderDid,
    sequenceNumber,
    sessionId,
  };
  // Sort keys for determinism (same as Python sort_keys=True)
  const sortedKeys = Object.keys(aadObj).sort() as Array<keyof typeof aadObj>;
  const sorted: Record<string, unknown> = {};
  for (const k of sortedKeys) {
    sorted[k] = aadObj[k];
  }
  return new TextEncoder().encode(JSON.stringify(sorted));
}

/**
 * Encrypt plaintext with AES-256-GCM and return a compact JWE-like string.
 *
 * Format: <protected>.<iv>.<ciphertext>.<tag>
 */
export async function encryptPayload(
  plaintext: Uint8Array,
  key: Uint8Array,
  sessionId: string,
  sequenceNumber: number,
  senderDid: string
): Promise<string> {
  if (key.length !== 32) {
    throw new Error("key must be 32 bytes for AES-256-GCM");
  }

  const iv = webcrypto.getRandomValues(new Uint8Array(12)); // 96-bit nonce
  const aad = buildAad(sessionId, sequenceNumber, senderDid);

  const cryptoKey = await webcrypto.subtle.importKey(
    "raw",
    key,
    { name: "AES-GCM" },
    false,
    ["encrypt"]
  );

  const ciphertextWithTag = await webcrypto.subtle.encrypt(
    { name: "AES-GCM", iv, additionalData: aad, tagLength: 128 },
    cryptoKey,
    plaintext
  );

  const ctBytes = new Uint8Array(ciphertextWithTag);
  const ciphertext = ctBytes.slice(0, ctBytes.length - 16);
  const tag = ctBytes.slice(ctBytes.length - 16);

  const protectedHeader = {
    alg: "dir",
    apu: b64urlEncode(aad),
    enc: "A256GCM",
  };
  const protectedB64 = b64urlEncode(
    new TextEncoder().encode(JSON.stringify(protectedHeader))
  );
  const ivB64 = b64urlEncode(iv);
  const ciphertextB64 = b64urlEncode(ciphertext);
  const tagB64 = b64urlEncode(tag);

  return `${protectedB64}.${ivB64}.${ciphertextB64}.${tagB64}`;
}

/**
 * Decrypt a compact JWE-like string produced by encryptPayload.
 */
export async function decryptPayload(
  jwe: string,
  key: Uint8Array,
  sessionId: string,
  sequenceNumber: number,
  senderDid: string
): Promise<Uint8Array> {
  if (key.length !== 32) {
    throw new Error("key must be 32 bytes for AES-256-GCM");
  }

  const parts = jwe.split(".");
  if (parts.length !== 4) {
    throw new Error("Invalid JWE: expected 4 dot-separated parts");
  }

  const [, ivB64, ciphertextB64, tagB64] = parts;
  const iv = b64urlDecode(ivB64);
  const ciphertext = b64urlDecode(ciphertextB64);
  const tag = b64urlDecode(tagB64);

  const aad = buildAad(sessionId, sequenceNumber, senderDid);

  const cryptoKey = await webcrypto.subtle.importKey(
    "raw",
    key,
    { name: "AES-GCM" },
    false,
    ["decrypt"]
  );

  // Concatenate ciphertext + tag for AES-GCM decrypt
  const ctWithTag = new Uint8Array(ciphertext.length + tag.length);
  ctWithTag.set(ciphertext, 0);
  ctWithTag.set(tag, ciphertext.length);

  try {
    const plaintext = await webcrypto.subtle.decrypt(
      { name: "AES-GCM", iv, additionalData: aad, tagLength: 128 },
      cryptoKey,
      ctWithTag
    );
    return new Uint8Array(plaintext);
  } catch (err) {
    throw new Error(`Decryption failed: ${err}`);
  }
}

File: sdap-python/src/sdap/crypto/encryption.py
"""AES-256-GCM encryption/decryption for SDAP session payloads (JWE-like format)."""

from __future__ import annotations

import base64
import json
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _b64url_decode(s: str) -> bytes:
    padding = 4 - len(s) % 4
    if padding != 4:
        s += "=" * padding
    return base64.urlsafe_b64decode(s)


def _build_aad(session_id: str, sequence_number: int, sender_did: str) -> bytes:
    """Build Additional Authenticated Data bytes."""
    aad_obj = {
        "sessionId": session_id,
        "sequenceNumber": sequence_number,
        "senderDID": sender_did,
    }
    return json.dumps(aad_obj, separators=(",", ":"), sort_keys=True).encode("utf-8")


def encrypt_payload(
    plaintext: bytes,
    key: bytes,
    session_id: str,
    sequence_number: int,
    sender_did: str,
) -> str:
    """Encrypt plaintext with AES-256-GCM and return a compact JWE-like string.

    Format: ``<protected>.<iv>.<ciphertext>.<tag>``

    The ``<protected>`` header contains algorithm metadata.
    AAD is derived from sessionId, sequenceNumber, and senderDID.
    """
    if len(key) != 32:
        raise ValueError("key must be 32 bytes for AES-256-GCM")

    iv = os.urandom(12)  # 96-bit nonce
    aad = _build_aad(session_id, sequence_number, sender_did)

    aesgcm = AESGCM(key)
    # AESGCM.encrypt returns ciphertext + 16-byte tag appended
    ciphertext_with_tag = aesgcm.encrypt(iv, plaintext, aad)
    ciphertext = ciphertext_with_tag[:-16]
    tag = ciphertext_with_tag[-16:]

    protected_header = {
        "alg": "dir",
        "enc": "A256GCM",
        "apu": _b64url_encode(aad),
    }
    protected_b64 = _b64url_encode(
        json.dumps(protected_header, separators=(",", ":")).encode()
    )
    iv_b64 = _b64url_encode(iv)
    ciphertext_b64 = _b64url_encode(ciphertext)
    tag_b64 = _b64url_encode(tag)

    return f"{protected_b64}.{iv_b64}.{ciphertext_b64}.{tag_b64}"


def decrypt_payload(
    jwe: str,
    key: bytes,
    session_id: str,
    sequence_number: int,
    sender_did: str,
) -> bytes:
    """Decrypt a compact JWE-like string produced by :func:`encrypt_payload`.

    Raises ``ValueError`` if the AAD doesn't match or decryption fails.
    """
    if len(key) != 32:
        raise ValueError("key must be 32 bytes for AES-256-GCM")

    parts = jwe.split(".")
    if len(parts) != 4:
        raise ValueError("Invalid JWE: expected 4 dot-separated parts")

    _, iv_b64, ciphertext_b64, tag_b64 = parts
    iv = _b64url_decode(iv_b64)
    ciphertext = _b64url_decode(ciphertext_b64)
    tag = _b64url_decode(tag_b64)

    aad = _build_aad(session_id, sequence_number, sender_did)
    aesgcm = AESGCM(key)
    try:
        plaintext = aesgcm.decrypt(iv, ciphertext + tag, aad)
    except Exception as exc:
        raise ValueError(f"Decryption failed: {exc}") from exc

    return plaintext
